raphtory/src/serialise/parquet/mod.rs

```diff
@@ -49,18 +55,115 @@ mod nodes;
 mod graph;
 
 pub trait ParquetEncoder {
+    fn encode_parquet_to_bytes(&self) -> Result<Vec<u8>, GraphError> {
+        // Write directly to an in-memory cursor
+        let mut zip_buffer = Vec::new();
+        let cursor = std::io::Cursor::new(&mut zip_buffer);
+
+        self.encode_parquet_to_zip(cursor)?;
+
+        Ok(zip_buffer)
+    }
+
+    fn encode_parquet_to_zip<W: Write + Seek>(&self, writer: W) -> Result<(), GraphError> {
+        // Encode to a tmp dir using parquet, then zip it to the writer
+        let temp_dir = tempfile::tempdir()?;
+        self.encode_parquet(&temp_dir)?;
+
+        let mut zip_writer = ZipWriter::new(writer);
+
+        // Walk through the directory and add files and directories to the zip.
+        // Files and directories are stored in the archive under the GRAPH_PATH directory.
+        for entry in WalkDir::new(temp_dir.path())
+            .into_iter()
+            .filter_map(Result::ok)
+        {
+            let path = entry.path();
+
+            let relative_path = path.strip_prefix(temp_dir.path()).map_err(|e| {
+                GraphError::IOErrorMsg(format!("Failed to strip prefix from path: {}", e))
+            })?;
+
+            // Attach GRAPH_PATH as a prefix to the relative path
+            let zip_entry_name = PathBuf::from(GRAPH_PATH)
+                .join(relative_path)
+                .to_string_lossy()
+                .into_owned();
+
+            if path.is_file() {
+                zip_writer.start_file::<_, ()>(zip_entry_name, FileOptions::<()>::default())?;
+
+                let mut file = std::fs::File::open(path)?;
+                std::io::copy(&mut file, &mut zip_writer)?;
+            } else if path.is_dir() {
+                // Add empty directories to the zip
+                zip_writer.add_directory::<_, ()>(zip_entry_name, FileOptions::<()>::default())?;
+            }
+        }
+
+        zip_writer.finish()?;
+        Ok(())
+    }
+
     fn encode_parquet(&self, path: impl AsRef<Path>) -> Result<(), GraphError>;
 }
 
-pub trait ParquetDecoder {
-    fn decode_parquet(path: impl AsRef<Path>) -> Result<Self, GraphError>
-    where
-        Self: Sized;
+pub trait ParquetDecoder: Sized {
+    fn decode_parquet_from_bytes(bytes: &[u8], path_for_decoded_graph: Option<impl AsRef<Path>>) -> Result<Self, GraphError> {
+        // Read directly from an in-memory cursor
+        let reader = std::io::Cursor::new(bytes);
+
+        Self::decode_parquet_from_zip(reader, path_for_decoded_graph)
+    }
+
+    fn decode_parquet_from_zip<R: Read + Seek>(reader: R, path_for_decoded_graph: Option<impl AsRef<Path>>) -> Result<Self, GraphError> {
+        // Unzip to a temp dir and decode parquet from there
+        let mut zip = zip::ZipArchive::new(reader)?;
```

Comment:
I wish we didn't have to write the actual files but we can try to revisit that later.

Reply:
we would have to change parquet encoding to natively encode in-memory, which may not be too hard but probably not for this PR...